src/PDDoc.jl

```diff
@@ -181,3 +182,55 @@ function pdDocGetNamesDict(doc::PDDoc)
     ref = get(catalog, CosName("Names"))
     obj = cosDocGetObject(doc.cosDoc, ref)
 end
+
+"""
+```
+    pdDocGetOutline(doc::PDDoc; depth::Number = Inf, compact::Bool = false, add_index::Bool = false) -> PDOutline
+```
+Given a PDF document provides the document Outline (Table of Contents)
+available in the `Document Catalog` dictionary.
+Returned object is an array of either `PDOutlineItem` or `PDOutline` items.
+Nested `PDOutline` is allways preceded by `PDOutlineItem` and contain its sub-sections.
+First element of `PDOutline` array is allways `PDOutlineItem`.
+
+`PDOutlineItem` represent outline item and is a dictionairy object with following keys:
+- `:Title` - outline displayed title (`CDTextString`), always present
+- `:Level` - nesting level of the item (`Int`), optional, not present in level 1 items or if `:Index` key is present
+- `:Index` - inxedes of the item in Outline structure (`NTuple`), optional, not present if `:Level` key is present.
+             If index is (a,b,c) the item can be referred as outline[a][b][c], where outline is a object returned by this function.
+- `:Expanded` - weather child items should be expanded by the GUI viewer by default (`Bool`), optional
+- `:Style` - the style which should be applied to item by the GUI viewer (`Int`) - refer to PDF specification, optional.
+- `:PageRef` - indirect reference to respective page (`CosIndirectObjectRef`) - use with other API functions eg. `cosDocGetObject`, present if not in compact mode
+- `:PageNo` - absulute number of respective page (`Int`), present if not in compact mode
+- `:PageLabel` - displayable label of respective page (`LabelNumeral` - can be casted to `String`), optional.
+
+Optional, named parameters:
+- `depth` - limits retrieved items to certain nesting level (0 for root chapters), default: no limit
+- `compact` - if `true`, only `:Title` and `:Level` or `:Index` are retrived - works substantially faster, default: `false`
+- `add_index` - if `true`, `:Index` key is placed instead of `:Level` - with minimal overhead, default: `false`
+
+If document does not have Outline, this method returns `nothing`.
+
+Note: This method extracts most important information from PDF Outline entry.
+There are more information stored in those PDF structures. Use low level functions of this library to extract them if necessary.
+"""
+function pdDocGetOutline(doc::PDDoc;
+        depth::Number = Inf,
```

Comment:
Function signature looks way to complicated. Ideally, APIs should be granular and should only do a simple task. All that I would expect in such a method will be to a "Representative text" and a "Destination" mapping for each outline item. And a parent child tree which can be enumerated through the data structure. Compact, depth, add_index etc are only making the whole API far more complex than is required. If such data can be obtained they can be properties of outline items.

Reply:
Note that those additional parameters have sensible defaults and to use the function you do not need to know about them at all. I think that 80% or more users will be happy with simple `pdDocGetOutline(doc)` call and don't need to bother with other options at all.
Additional parameters are only in case if you want to slightly modify function behavior, and every one has proper rationale:
- `compact` - option compact=false require a lot more processing and reading additional PDF structures then compact option, which more-less reads only Outlines catalog entry. Form the other side, there are use cases where user may only be interested in Outline contents - not the page numbers or references Generally there are two possible library usage scenarios: when you work with one or few documents and need more complete information vs. you do some very massive file scanning against millions of documents and need only specific information and speed (as well as reliability, memory etc.) is your concern.
- `depth` - similar reasons as above - when doing massive scanning / indexing you may be interested only with main chapters and sections without going into too much details, but again want to do scan as fast as possible and allocate as less memory as possible.
- `add_index` - little more tricky. At first I implemented the flag passed to `show() ` to display or not indexes and did not have any explicit level information inside the item. But later found it  quite inconvenient because `show` function is being called automatically by Julia which does not know about this parameter so it was usually lost. Also later having implemented an iterator, found it help-full to have an `item_level` function available for every item, so changed design to remember level information inside the item.
Why it is not always `:Level`? - Because with longer documents with many items it is helfull to be able to display indexes.
Why it is not always `:Index`? - Because in 90% cases `:Level` is enough and is lighter, and also because I would like to be able to directly map returned object to e.g. JSON and have verbose (not calculated) information about level.

Indeed for application than I'm writing (massive PDF scan / indexing) I strictly need compact=true and depth=1 options (i.e. some not defaults), but I understand that there are very different scenarios also, and users when calling functions with defaults usually need complete information rather then performance.

Julia although very different internally but is quite similar to Python on surface and to some degree share the same user base. Such optional, named parameters are in common use in all Python standard and add-on libraries and does very good job making code maybe little longer but a lot  cleaner and easier to understand without poking documentation all the time.